### PR TITLE
Coordinated-fanout

### PR DESCRIPTION
**Pull request by niteshbalusu11**
Signed-off-by: Nitesh Balusu <niteshbalusu@icloud.com>

~~I need some help, the final signed tx is not being accepted by bitcoin-core. i am able to decode the constructed tx and it shows the inputs and outputs though.~~

Ready for review.

2 new bos commands:
- `create-fanout-group` and `join-fanout-group`
- For given inputs (can pass your own) it generates equal amount outputs of the size you want.
- Min group size is 3
- The coordinator decides the output size for all peers, joiners can decide how many outputs they want.
- Default output count is 1 excluding change

**Important note: this is a breaking change for group-open commands.**
The tlv encoding and decoding for funding scripts has changed which would break group open. So the version number has been bumped.

### CODE REVIEW

**Comment by alexbosworth** [comment]
seems like it needs integration test, similar to the group ones

**Comment by alexbosworth** [comment]
also i think it may break some existing integration tests

**Comment by niteshbalusu11** [comment]
All the existing integration tests are passing. Can make new ones for this.

**Comment by niteshbalusu11** [comment]
Added a integration test.

**Comment by alexbosworth** [comment]
why does it put so many inputs in the fanout tx?

**Comment by alexbosworth** [comment]
my guess is what's going on is that if you don't select inputs it is putting all the inputs in which is why i see a change output that has thousands of BTC in the integration test, so I think it's simple to just always require that all participants always explicitly select all UTXOs

**Comment by niteshbalusu11** [comment]
Should be fixed now.

**Comment by niteshbalusu11** [comment]
Utxos selection is now required. Updated the integration test.

File: fanout/assemble_fanout_group.js [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+const times = 2 * 60 * 10;
+
+/** Assemble a fanout group
+
+  {
+    capacity: <Output Capacity Tokens Number>
+    count: <Channel Members Count Number>
+    ecp: <ECPair Library Object>
+    identity: <Coordinator Identity Public Key Hex String>
+    [inputs]: [<Utxo Outpoint String>]
```
Comment (alexbosworth) [inline fanout/assemble_fanout_group.js]: This should be required, to say which inputs you want to be paid into the fanout

File: fanout/assemble_fanout_group.js [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    rate: <Chain Fee Tokens Per VByte Number>
+  }
+
+  @returns
+  {
+    events: <EventEmitter Object>
+    id: <Group Id Hex String>
+  }
+
+  // Open was published
```
Comment (alexbosworth) [inline fanout/assemble_fanout_group.js]: There isn't an open in a fanout

File: fanout/assemble_fanout_group.js [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+  // An error was encountered
+  coordinator.events.once('error', errored);
+
+  // Group members have registered themselves
+  coordinator.events.once('joined', async ({ids}) => {
+    emitter.emit('filled', {ids});
+
+    try {
+      const {inbound, outbound} = coordinator.partners(args.identity);
```
Comment (alexbosworth) [inline fanout/assemble_fanout_group.js]: For a fanout there isn't really a need for partners

File: fanout/assemble_fanout_group.js [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+      // Connect to the inbound and outbound partners
+      await peerWithPartners({
```
Comment (alexbosworth) [inline fanout/assemble_fanout_group.js]: peering with partners isn't needed

File: fanout/coordinate_fanout.js [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+const {serviceTypeConfirmConnected} = require('./../service_types');
+const {serviceTypeFindFanoutPartners} = require('./../service_types');
+const {serviceTypeGetFanoutDetails} = require('./../service_types');
+const {serviceTypeRegisterPendingFanout} = require('./../service_types');
+const {serviceTypeRegisterSignedFanout} = require('./../service_types');
+
+const findRecord = (records, type) => records.find(n => n.type === type);
+const {isArray} = Array;
+const makeGroupId = () => randomBytes(16).toString('hex');
+const minGroupCount = 2;
```
Comment (alexbosworth) [inline fanout/coordinate_fanout.js]: is the min group size 2 or 3?

**Comment by niteshbalusu11** [comment]
Ok should be good now. Peering step has been removed.

File: fanout/assemble_fanout_group.js [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+      pending.utxos = utxos;
+
+      // Register as proposed
+      return coordinator.proposed({change, funding, utxos, id: args.identity});
+    } catch (err) {
+      return errored(err);
+    }
+  });
+
+  // Group members have proposed channels to each other
```
Comment (alexbosworth) [inline fanout/assemble_fanout_group.js]: there aren't channels right

File: fanout/create_group_fanout.js [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+          return cbk(null, args.inputs);
+        }
+
+        // Exit early when not selecting UTXOs interactively
+        if (!args.is_selecting_utxos) {
+          return cbk(null, []);
+        }
+
+        // Only selecting confirmed utxos is supported
+        const utxos = getUtxos.utxos.filter(n => !!n.confirmation_count);
```
Comment (alexbosworth) [inline fanout/create_group_fanout.js]: i think you also want to filter out np2wpkh

File: fanout/create_group_fanout.js [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+          const nodes = await asyncMap(members, async id => {
+            return niceName(await getNodeAlias({id, lnd: args.lnd}));
+          });
+
+          return args.logger.info({ready: join(nodes)});
+        });
+
+        // Once filled, members will connect with their partners
+        coordinate.events.once('connected', () => {
+          return args.logger.info({peered: true});
```
Comment (alexbosworth) [inline fanout/create_group_fanout.js]: why do we need a peered step?

File: fanout/propose_fanout.js [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+const tinysecp = require('tiny-secp256k1');
+
+const asOutpoint = n => `${n.transaction_id}:${n.transaction_vout}`;
+const {isArray} = Array;
+const isEven = n => !(n % 2);
+const isNumber = n => !isNaN(n);
+const nestedSegWitAddressFormat = 'np2wpkh';
+const nestedSegWitPath = "m/49'/";
+
+/** Fund and propose a fanout to a peer
```
Comment (alexbosworth) [inline fanout/propose_fanout.js]: This isn't really named correctly, you aren't proposing a fanout to a peer you are just creating fanout funding output scripts here

File: fanout/propose_fanout.js [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+          if (err) {
+            return cbk(err);
+          }
+          
+          return cbk(null, {pending: res});
+        });
+      }],
+
+      // Lock nested segwit inputs
+      lockUtxos: ['getFilteredInputs', ({getFilteredInputs}, cbk) => {
```
Comment (alexbosworth) [inline fanout/propose_fanout.js]: since the inputs are defined i don't think this locking does anything

File: groups/assemble/assemble_unsigned_psbt.js [hunk 11]
```diff
@@ -83,9 +83,15 @@ module.exports = ({capacity, proposed, rate}, cbk) => {
...
           const inputsOffset = tx.ins.length;
 
-          [member.change, member.funding].filter(n => !!n).forEach(out => {
-            return tx.addOutput(hexAsBuffer(out), capacity);
-          });
+          if (!!member.change) {
+            tx.addOutput(hexAsBuffer(member.change), capacity);
+          }
+
+          if (!!member.funding && !!member.funding.length) {
```
Comment (alexbosworth) [inline groups/assemble/assemble_unsigned_psbt.js:90]: is there a case when there isn't member funding? that seems like a problem if you are just making change

File: groups/assemble/assemble_unsigned_psbt.js [hunk 12]
```diff
@@ -114,16 +120,29 @@ module.exports = ({capacity, proposed, rate}, cbk) => {
...
-            },
-            {
-              script: member.change,
-              tokens: funded - committed(capacity, proposed) - (vbytes * rate),
-            },
-          ];
+          const fee = vbytes * rate;
+
+          // Calculate total output amount (including funding outputs)
+          const memberlength = !!member.funding ? (member.funding.length - 1) : 0;
```
Comment (alexbosworth) [inline groups/assemble/assemble_unsigned_psbt.js]: i'm not sure what's going on here, maybe the comments or code could be a bit more obvious like why it's subtracting 1 and sometimes the member length is zero

File: groups/p2p/peer_with_partners.js [hunk 13]
```diff
@@ -12,26 +12,27 @@ const times = 2 * 60 * 30;
   {
     capacity: <Channel Capacity Tokens Number>
     [inbound]: <Inbound Identity Public Key Hex String>
+    [skip_acceptance_check]: <Skip Channel Acceptance Check Bool>
```
Comment (alexbosworth) [inline groups/p2p/peer_with_partners.js]: why is peering needed at all

File: service_types.json [hunk 14]
```diff
@@ -27,6 +30,9 @@
     "8050011": "serviceTypeFindGroupPartners",
     "8050012": "serviceTypeConfirmConnected",
     "8050013": "serviceTypeRegisterPendingOpen",
-    "8050014": "serviceTypeRegisterSignedOpen"
+    "8050014": "serviceTypeRegisterSignedOpen",
+    "8050016": "serviceTypeGetFanoutDetails",
```
Comment (alexbosworth) [inline service_types.json]: what happened to 15

File: test/integration/test_fanout.js [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+
+        const fanoutUtxosCount = utxos.filter(n => n.tokens === capacity).length;
+
+        if (outputCount !== fanoutUtxosCount) {
+          throw new Error('ExpectedCorrectFanoutUtxoCount');
+        }
+      }));
+    });
+
+    console.log('reached here 8');
```
Comment (alexbosworth) [inline test/integration/test_fanout.js]: console statement

File: groups/assemble/assemble_unsigned_psbt.js (hunk 11 above)
Comment (niteshbalusu11) [reply groups/assemble/assemble_unsigned_psbt.js:90]: that check was already sort of there before for group open where we filter out null values in the array. so i am just checking the same thing.

https://github.com/alexbosworth/paid-services/blob/master/groups/assemble/assemble_unsigned_psbt.js#L86

Comment (niteshbalusu11) [reply groups/assemble/assemble_unsigned_psbt.js:90]: and i think it is undefined in some case for group_open. i am not sure what the case is. maybe for pair groups.

File: fanout/join_group_fanout.js [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+          lnd: args.lnd,
+          output_count: args.output_count,
+          rate: getJoinDetails.rate,
+        });
+
+        join.once('end', ({id}) => cbk(null, {transaction_id: id}));
+        join.once('error', err => cbk(err));
+
+        // After the group is filled the members are matched and peer up
+        join.once('peering', async ({inbound, outbound}) => {
```
Comment (alexbosworth) [inline fanout/join_group_fanout.js]: when i go to joinFanout I don't see peering event existing, is this still actually happening?

Comment (niteshbalusu11) [reply fanout/join_group_fanout.js]: Hmm, maybe I missed removing this. 

No, there isn't any peering happening other than joiners peering with the coordinator.

**Comment by niteshbalusu11** [comment]
Fixed it.

File: fanout/create_group_fanout.js [hunk 17]
```diff
@@ -0,0 +1,325 @@
...
+const descriptionForGroup = 'group fanout';
+const asBigUnit = n => (n / 1e8).toFixed(8);
+const asOutpoint = utxo => `${utxo.transaction_id}:${utxo.transaction_vout}`;
+const {isArray} = Array;
+const isNumber = n => !isNaN(n);
+const isPublicKey = n => !!n && /^0[2-3][0-9A-F]{64}$/i.test(n);
+const isOdd = n => !!(n % 2);
+const isValidMembersCount = (n, count) => !n.length || n.length === count - 1;
+const join = arr => arr.join(', ');
+const maxGroupSize = 420;
```
Comment (alexbosworth) [inline fanout/create_group_fanout.js]: I think this should be lowered because you can have multiple outputs per user

How about 42 max?

File: fanout/join_group_fanout.js [hunk 18]
```diff
@@ -0,0 +1,202 @@
...
+
+        // Exit early when not selecting UTXOs interactively
+        if (!args.is_selecting_utxos) {
+          return cbk(null, []);
+        }
+
+        // Only selecting confirmed utxos is supported
+        const utxos = getUtxos.utxos
+        .filter(n => !!n.confirmation_count)
+        .filter(n => n.address_format !== nestedSegWitAddressFormat);
```
Comment (alexbosworth) [inline fanout/join_group_fanout.js]: I think this should really check for whitelisted output types to be safe against future output types that are unknown and therefore won't be supported, so like p2tr or p2wpkh

**Comment by niteshbalusu11** [comment]
Fixed max size and utxo check.

File: fanout/get_funding_details.js [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+          outputs: propose.pending.map(output => ({
+            address: output.address,
+            tokens: output.tokens,
+          })),
+        },
+        cbk);
+      }],
+
+      // Unlock UTXOs
+      unlock: ['fund', ({fund}, cbk) => {
```
Comment (alexbosworth) [inline fanout/get_funding_details.js:173]: why are the utxos being unlocked?

File: fanout/create_group_fanout.js [hunk 20]
```diff
@@ -0,0 +1,326 @@
...
+const assembleFanoutGroup = require('./assemble_fanout_group');
+
+const allowedAddressFormats = ['p2tr', 'p2wpkh'];
+const descriptionForGroup = 'group fanout';
+const asBigUnit = n => (n / 1e8).toFixed(8);
+const asOutpoint = utxo => `${utxo.transaction_id}:${utxo.transaction_vout}`;
+const {isArray} = Array;
+const isNumber = n => !isNaN(n);
+const isPublicKey = n => !!n && /^0[2-3][0-9A-F]{64}$/i.test(n);
+const isOdd = n => !!(n % 2);
```
Comment (alexbosworth) [inline fanout/create_group_fanout.js]: why do we care about odd? I think in group this is done to split balances 50/50

File: fanout/join_group_fanout.js [hunk 21]
```diff
@@ -0,0 +1,203 @@
+const asyncAuto = require('async/auto');
+const asyncMap = require('async/map');
+const {getMethods} = require('ln-service');
+const {getUtxos} = require('ln-service');
+const {getNodeAlias} = require('ln-sync');
```
Comment (alexbosworth) [inline fanout/join_group_fanout.js]: i don't think this is used anymore

File: groups/assemble/assemble_unsigned_psbt.js [hunk 22]
```diff
@@ -114,16 +120,40 @@ module.exports = ({capacity, proposed, rate}, cbk) => {
...
-              tokens: funded - committed(capacity, proposed) - (vbytes * rate),
-            },
-          ];
+          const fee = vbytes * rate;
+
+          // Calculate total output amount (including funding outputs)
+          // Group channel opens have only one funding output
+          // Fanouts have multiple funding outputs
+          const calculateTotalOutputAmount = (member, capacity, proposed) => {
+            const memberCount = !!member.funding ? member.funding.length : 0;
```
Comment (alexbosworth) [inline groups/assemble/assemble_unsigned_psbt.js:129]: when is there zero members? is this for like a pair group? it seems confusing to have zero members

Comment (niteshbalusu11) [reply groups/assemble/assemble_unsigned_psbt.js:129]: Yes, there is a case where if it's a pair group, member.funding is undefined.

File: fanout/get_funding_details.js [hunk 23]
```diff
@@ -0,0 +1,227 @@
...
+          lnd: args.lnd,
+          outputs: propose.pending.map(output => ({
+            address: output.address,
+            tokens: output.tokens,
+          })),
+        },
+        cbk);
+      }],
+
+      // Unlock UTXOs
```
Comment (alexbosworth) [inline fanout/get_funding_details.js]:
```suggestion
      // Unlock UTXOs to undo the PSBT funding locks
```

File: fanout/join_fanout.js [hunk 24]
```diff
@@ -0,0 +1,183 @@
...
+    {
+      const funding = hexAsBuffer(getFundingInfo.funding);
+      const psbt = decodePsbt({ecp, psbt: register.psbt});
+
+      const tx = fromHex(psbt.unsigned_transaction);
+
+      return cbk(null, {
+        id: tx.getId(),
+        raw: psbt.unsigned_transaction,
+        vout: tx.outs.findIndex(n => n.script.equals(funding)),
```
Comment (alexbosworth) [inline fanout/join_fanout.js]: What is vout used for?

File: fanout/join_fanout.js [hunk 25]
```diff
@@ -0,0 +1,183 @@
...
+    }],
+
+    // Decode the unsigned PSBT
+    transaction: [
+      'ecp',
+      'getFundingInfo',
+      'register',
+      ({ecp, getFundingInfo, register}, cbk) =>
+    {
+      const funding = hexAsBuffer(getFundingInfo.funding);
```
Comment (alexbosworth) [inline fanout/join_fanout.js]: isn't funding here an array not a hex?

File: fanout/join_fanout.js (hunk 24 above)
Comment (niteshbalusu11) [reply fanout/join_fanout.js]: actually this whole function can be removed. it's not used for anything.

File: fanout/join_fanout.js (hunk 25 above)
Comment (niteshbalusu11) [reply fanout/join_fanout.js]: actually this whole function can be removed. it's not used for anything.

File: fanout/join_fanout.js (hunk 24 above)
Comment (niteshbalusu11) [reply fanout/join_fanout.js]: actually nvm. should just remove vouts and funding

File: fanout/join_fanout.js (hunk 25 above)
Comment (niteshbalusu11) [reply fanout/join_fanout.js]: actually nvm. should just remove vouts and funding

File: groups/p2p/register_group_connected.js [hunk 26]
```diff
@@ -87,6 +87,10 @@ module.exports = ({coordinator, count, id, lnd}, cbk) => {
               return cbk(err);
             }
 
+            if (!res.records || !res.records.length) {
```
Comment (alexbosworth) [inline groups/p2p/register_group_connected.js:90]: is there a reason this is changed?

Comment (niteshbalusu11) [reply groups/p2p/register_group_connected.js:90]: It is because fanout has one less step so by the time it hits this step, records are an empty object so the decode will fail and it will error out until peers connect to the coordinator. 

This shouldn't effect group open because it will always have records at this point and it will hit decode step as usual.

File: groups/p2p/peer_with_partners.js [hunk 27]
```diff
@@ -18,20 +18,20 @@ const times = 2 * 60 * 30;
 
   @returns via cbk or Promise
 */
-module.exports = ({capacity, inbound, lnd, outbound}, cbk) => {
+module.exports = (args, cbk) => {
```
Comment (alexbosworth) [inline groups/p2p/peer_with_partners.js]: why is this changed now? is it still referenced?

Comment (niteshbalusu11) [reply groups/p2p/peer_with_partners.js]: Hmm, no. I referenced it before. Maybe I can just revert the file to the master branch.

**Comment by niteshbalusu11** [comment]
Fixed.

File: fanout/join_group_fanout.js [hunk 28]
```diff
@@ -0,0 +1,202 @@
...
+const {getUtxos} = require('ln-service');
+const {returnResult} = require('asyncjs-util');
+
+const joinFanout = require('./join_fanout');
+const getJoinDetails = require('./get_join_details');
+
+const allowedAddressFormats = ['p2tr', 'p2wpkh'];
+const asBigUnit = n => (n / 1e8).toFixed(8);
+const asOutpoint = utxo => `${utxo.transaction_id}:${utxo.transaction_vout}`;
+const formatNodes = arr => arr.join(', ');
```
Comment (alexbosworth) [inline fanout/join_group_fanout.js]: I think this can be removed, it's not used?

File: fanout/join_group_fanout.js [hunk 29]
```diff
@@ -0,0 +1,202 @@
...
+const {isArray} = Array;
+const isCode = n => !!n && n.length === 98;
+const isNumber = n => !isNaN(n);
+const nestedSegWitAddressFormat = 'np2wpkh';
```
Comment (alexbosworth) [inline fanout/join_group_fanout.js]: is this still used now?

File: fanout/join_group_fanout.js [hunk 30]
```diff
@@ -0,0 +1,202 @@
...
+const niceName = ({alias, id}) => `${alias} ${id}`.trim();
```
Comment (alexbosworth) [inline fanout/join_group_fanout.js]: used?

File: fanout/assemble_fanout_group.js [hunk 31]
```diff
@@ -0,0 +1,183 @@
...
+  // Group members have submitted their partial signatures
+  coordinator.events.once('signed', async () => {
+    // Collect all the partially signed PSBTs
+    const psbts = coordinator.signed().map(n => n.signed);
+
+    emitter.emit('signed', {psbts});
+
+    try {
+      // Merge partial PSBTs into a single PSBT
+      const combined = combinePsbts({ecp: args.ecp, psbts});
```
Comment (alexbosworth) [inline fanout/assemble_fanout_group.js]:
```suggestion
      const combined = combinePsbts({psbts, ecp: args.ecp});
```

**Comment by niteshbalusu11** [comment]
Removed unused stuff in all fanout files.

File: fanout/get_funding_details.js [hunk 32]
```diff
@@ -0,0 +1,227 @@
...
+      }],
+
+      // Generate addresses to fund
+      propose: ['validate', ({}, cbk) => {
+        return asyncMap(Array(args.output_count).fill(), (_, cbk) => {
+          return createChainAddress({is_unused: false, lnd: args.lnd}, (err, res) => {
+            if (err) {
+              return cbk(err);
+            }
+
```
Comment (alexbosworth) [inline fanout/get_funding_details.js]:
```suggestion

```

File: fanout/get_funding_details.js [hunk 33]
```diff
@@ -0,0 +1,227 @@
...
+              return cbk(err);
+            }
+            
+            return cbk(null, {address: res.address, tokens: args.capacity});
+          });
+        }, (err, res) => {
+          if (err) {
+            return cbk(err);
+          }
+
```
Comment (alexbosworth) [inline fanout/get_funding_details.js]:
```suggestion
          
```

File: fanout/get_funding_details.js [hunk 34]
```diff
@@ -0,0 +1,227 @@
...
+            return cbk(err);
+          }
+          
+          return cbk(null, {pending: res});
+        });
+      }],
+
+      // Fund the address to populate UTXOs that can be used
+      fund: [
+      'getFilteredInputs',
```
Comment (alexbosworth) [inline fanout/get_funding_details.js]:
```suggestion
        'getFilteredInputs',
```

File: fanout/get_funding_details.js [hunk 35]
```diff
@@ -0,0 +1,227 @@
...
+      'propose',
```
Comment (alexbosworth) [inline fanout/get_funding_details.js]:
```suggestion
        'propose',
```

File: fanout/join_fanout.js [hunk 36]
```diff
@@ -0,0 +1,177 @@
...
+    },
+
+    // Confirm connected the partners
+    connected: ['validate', ({}, cbk) => {
+      return registerGroupConnected({
+        coordinator: args.coordinator,
+        count: args.count,
+        id: args.id,
+        lnd: args.lnd,
+      }, cbk);
```
Comment (alexbosworth) [inline fanout/join_fanout.js]:
```suggestion
      },
      cbk);
```

File: fanout/join_fanout.js [hunk 37]
```diff
@@ -0,0 +1,177 @@
...
+      }
+
+      if (!args.rate) {
+        return cbk([400, 'ExpectedChainFeeRateToJoinFanout']);
+      }
+
+      return cbk();
+    },
+
+    // Confirm connected the partners
```
Comment (alexbosworth) [inline fanout/join_fanout.js]: what is happening in this step, considering that there are no partners?

Comment (niteshbalusu11) [reply fanout/join_fanout.js]: It's just a step to ensure everyone is connected to the coordinator.
